---
title: "Basic ChIP-seq Analysis (from BAM to Metaplot)"
output: 
    html_document:
        toc: TRUE
---


<style>
a {
  font-size: 16px;
}
</style>


```{r global options, include = FALSE}
knitr::opts_chunk$set(echo=TRUE, include = TRUE, warning=FALSE, message=FALSE)
```

---

## Embryo ChIP-seq Data

* MNase ChIP-seq data from 3-4 h Drosophila embryos
* MSL2 vs Input samples
* sub-nucleosomal fragments are considered (paired-end data)


---

### Pre-processing

* alignment was performed using bowtie2
* fragment length was set between 10 and 130

```{bash eval=F}
# do not run here
module load ngs/bowtie2/2.2.9
module load ngs/samtools/1.3.1

BOWTIE_INDEX="../Drosophila_melanogaster/UCSC/dm6/Sequence/Bowtie2Index/genome"

BOWTIE_OPTS="-p 24 --end-to-end --very-sensitive --no-unal --no-mixed --no-discordant -I 10 -X 130"
bowtie2 $BOWTIE_OPTS -x $BOWTIE_INDEX -1 ${FILEBASE}_1.txt.gz -2 ${FILEBASE}_2.txt.gz  > ${FILEBASE}.sam 2> ${FILEBASE}.stats
samtools view -bS -@ 24 -q 12  ${FILEBASE}.sam | samtools sort -@ 24 - | tee ${FILEBASE}.bam | samtools index - ${FILEBASE}.bam.bai
```

---

### Subsampling

* for this demo bam files were first subsampled to 1 million reads
* after subsampling only X chromosomal reads are considered

```{bash eval=F}
# do not run here
CHR='chrX'

TOTAL=`samtools view -c ${FILEBASE}.bam`
SCALER=`echo "scale=10; 1000000/(${TOTAL})" | bc`

samtools view -b -s ${SCALER} ${FILEBASE}.bam > ${FILEBASE}.1M.bam
samtools index ${FILEBASE}.1M.bam

samtools view -b ${FILEBASE}.1M.bam ${CHR} > ${FILEBASE}.1M.${CHR}.bam
samtools index ${FILEBASE}.1M.${CHR}.bam
```

---

# Metaplots

### Sites

* peaks were defined previously by Straub et al. 2013
* these are well characterized high affinity sites (HAS)
* almost all of these sites are on the X chromosome

```{r}
library(rtracklayer)

sites <- import.bed("data_BasicChIPSeq/hasC_peaks_dm6_ucsc.bed")

# take a window of 4000 surrounding the center
win_size <- 4000
sites_4000 <- resize(sites, width = win_size, fix = "center")

sites_4000
```

---

### Coverage

* iterate through BAM files to create coverage
* no additional normalization is needed as bam files were subsampled

```{r}
# function to create coverage from BAM files
library(GenomicAlignments)

bam2cov <- function(bam_file, type = c("SINGLE","PAIRED"), width = 150, MAPQ = 255){
    
    if(type == "SINGLE"){
        
        my_bam <- readGAlignments(bam_file, param = ScanBamParam(mapqFilter = MAPQ))
        grs <- as(my_bam, "GRanges")
        grsr <- resize(grs, width)
        
        cov <- coverage(grsr)
        return(cov)
        
    } else if(type == "PAIRED"){
        
        my_bam <- readGAlignmentPairs(bam_file,param = ScanBamParam(mapqFilter = MAPQ))
        grsr <- as(my_bam, "GRanges")
        
        cov <- coverage(grsr)
        return(cov)
    }
}
```


```{r results='hide'}
# set bam file path
data_dir <- "data_BasicChIPSeq/"
bam_files <- list.files(path = data_dir, pattern = ".bam$")
bam_path <- file.path(data_dir, bam_files)

# parallel processing
parallel::mclapply(seq_along(bam_files), mc.cores = 8, FUN = function(i){
    
    bam_name <- paste0("coverage.", gsub(".bam","",bam_files[i]))
    
    # create coverage
    bam_coverage <- bam2cov(bam_file = bam_path[i], 
                           type = "PAIRED", MAPQ = 12)
    
    assign(bam_name, bam_coverage)
    
    # save as rda file
    save(list = bam_name, file = paste0(data_dir, bam_name, ".rda"))
    
})

# load coverage files
cov_files <- file.path(data_dir, list.files(path = data_dir, pattern = ".rda$"))

for(i in seq_along(cov_files)){
    load(cov_files[i])
}

```

---

### Site Centered Matrix

* coverages (RleList) can be simply subset by windows surrounding sites


```{r fig.align="center", fig.height=8, fig.width=10}
covs <- ls(pattern = "coverage.")

par(mfcol=c(2,3), cex=1.1)

for(i in seq_along(covs)){
    
    covv <- get(covs[i])
    
    # format name
    main_title <- gsub(".*E03_|_[G,A,T,C][G,A,T,C].*", "", covs[i])
    
    plot(colMeans(as.matrix(covv[sites_4000])), 
         type="l", lwd=2, col = "red3", 
         ylim = c(0,10), xaxt="n",
         ylab = "Coverage", xlab = "Center",
         main = main_title)
    
    axis(side = 1, at = seq(1,win_size, length.out = 5), 
         labels =  c(-(win_size/2),"",0,"",(win_size/2))  )
    
    
}




```

---

### Input Normalization


---

# Data Source

The  dataset was generated by Nadia Prayitno (LMU, BMC, Becker group)


---


# References

Progressive dosage compensation during Drosophila embryogenesis is reflected by gene arrangement.
Prayitno K, Schauer T, Regnard C, Becker PB.
EMBO Rep. 2019 Aug;20(8):e48138. doi: 10.15252/embr.201948138. Epub 2019 Jul 9.
PMID: 31286660

Note: this pipeline (subsampling etc.) differs from the papers version, but findings are the same either way
