<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Basic ChIP-seq Analysis (from BAM to Metaplot)</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="styles.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 60px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h2 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h3 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h4 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h5 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h6 {
  padding-top: 65px;
  margin-top: -65px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Home</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="biostats.html">BioStats</a>
</li>
<li>
  <a href="genomics.html">Genomics</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Basic ChIP-seq Analysis (from BAM to Metaplot)</h1>

</div>

<div id="TOC">
<ul>
<li><a href="#embryo-chip-seq-data">Embryo ChIP-seq Data</a><ul>
<li><a href="#pre-processing">Pre-processing</a></li>
<li><a href="#subsampling">Subsampling</a></li>
</ul></li>
<li><a href="#metaplots">Metaplots</a><ul>
<li><a href="#sites">Sites</a></li>
<li><a href="#coverage">Coverage</a></li>
<li><a href="#site-centered-matrix">Site Centered Matrix</a></li>
<li><a href="#input-normalization">Input Normalization</a></li>
</ul></li>
<li><a href="#data-source">Data Source</a></li>
<li><a href="#references">References</a></li>
</ul>
</div>

<style>
a {
  font-size: 16px;
}
</style>
<hr />
<div id="embryo-chip-seq-data" class="section level2">
<h2>Embryo ChIP-seq Data</h2>
<ul>
<li>MNase ChIP-seq data from 3-4 h Drosophila embryos</li>
<li>MSL2 vs Input samples</li>
<li>sub-nucleosomal fragments are considered (paired-end data)</li>
</ul>
<hr />
<div id="pre-processing" class="section level3">
<h3>Pre-processing</h3>
<ul>
<li>alignment was performed using bowtie2</li>
<li>fragment length was set between 10 and 130</li>
</ul>
<pre class="bash"><code># do not run here
module load ngs/bowtie2/2.2.9
module load ngs/samtools/1.3.1

BOWTIE_INDEX=&quot;../Drosophila_melanogaster/UCSC/dm6/Sequence/Bowtie2Index/genome&quot;

BOWTIE_OPTS=&quot;-p 24 --end-to-end --very-sensitive --no-unal --no-mixed --no-discordant -I 10 -X 130&quot;
bowtie2 $BOWTIE_OPTS -x $BOWTIE_INDEX -1 ${FILEBASE}_1.txt.gz -2 ${FILEBASE}_2.txt.gz  &gt; ${FILEBASE}.sam 2&gt; ${FILEBASE}.stats
samtools view -bS -@ 24 -q 12  ${FILEBASE}.sam | samtools sort -@ 24 - | tee ${FILEBASE}.bam | samtools index - ${FILEBASE}.bam.bai</code></pre>
<hr />
</div>
<div id="subsampling" class="section level3">
<h3>Subsampling</h3>
<ul>
<li>for this demo bam files were first subsampled to 1 million reads</li>
<li>after subsampling only X chromosomal reads are considered</li>
</ul>
<pre class="bash"><code># do not run here
CHR=&#39;chrX&#39;

TOTAL=`samtools view -c ${FILEBASE}.bam`
SCALER=`echo &quot;scale=10; 1000000/(${TOTAL})&quot; | bc`

samtools view -b -s ${SCALER} ${FILEBASE}.bam &gt; ${FILEBASE}.1M.bam
samtools index ${FILEBASE}.1M.bam

samtools view -b ${FILEBASE}.1M.bam ${CHR} &gt; ${FILEBASE}.1M.${CHR}.bam
samtools index ${FILEBASE}.1M.${CHR}.bam</code></pre>
<hr />
</div>
</div>
<div id="metaplots" class="section level1">
<h1>Metaplots</h1>
<div id="sites" class="section level3">
<h3>Sites</h3>
<ul>
<li>peaks were defined previously by Straub et al. 2013</li>
<li>these are well characterized high affinity sites (HAS)</li>
<li>almost all of these sites are on the X chromosome</li>
</ul>
<pre class="r"><code>library(rtracklayer)

sites &lt;- import.bed(&quot;data_BasicChIPSeq/hasC_peaks_dm6_ucsc.bed&quot;)

# take a window of 4000 surrounding the center
win_size &lt;- 4000
sites_4000 &lt;- resize(sites, width = win_size, fix = &quot;center&quot;)

sites_4000</code></pre>
<pre><code>## GRanges object with 247 ranges and 0 metadata columns:
##         seqnames            ranges strand
##            &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt;
##     [1]     chrX 11578250-11582249      *
##     [2]     chrX 17819715-17823714      *
##     [3]     chrX 16796370-16800369      *
##     [4]     chrX   5073980-5077979      *
##     [5]     chrX 10863800-10867799      *
##     ...      ...               ...    ...
##   [243]     chrX 22459702-22463701      *
##   [244]     chrX 18633480-18637479      *
##   [245]     chrX   4925585-4929584      *
##   [246]     chrX   4630760-4634759      *
##   [247]     chrX   6534805-6538804      *
##   -------
##   seqinfo: 3 sequences from an unspecified genome; no seqlengths</code></pre>
<hr />
</div>
<div id="coverage" class="section level3">
<h3>Coverage</h3>
<ul>
<li>iterate through BAM files to create coverage</li>
<li>no additional normalization is needed as bam files were subsampled</li>
</ul>
<pre class="r"><code># function to create coverage from BAM files
library(GenomicAlignments)

bam2cov &lt;- function(bam_file, type = c(&quot;SINGLE&quot;,&quot;PAIRED&quot;), width = 150, MAPQ = 255){
    
    if(type == &quot;SINGLE&quot;){
        
        my_bam &lt;- readGAlignments(bam_file, param = ScanBamParam(mapqFilter = MAPQ))
        grs &lt;- as(my_bam, &quot;GRanges&quot;)
        grsr &lt;- resize(grs, width)
        
        cov &lt;- coverage(grsr)
        return(cov)
        
    } else if(type == &quot;PAIRED&quot;){
        
        my_bam &lt;- readGAlignmentPairs(bam_file,param = ScanBamParam(mapqFilter = MAPQ))
        grsr &lt;- as(my_bam, &quot;GRanges&quot;)
        
        cov &lt;- coverage(grsr)
        return(cov)
    }
}</code></pre>
<pre class="r"><code># set bam file path
data_dir &lt;- &quot;data_BasicChIPSeq/&quot;
bam_files &lt;- list.files(path = data_dir, pattern = &quot;.bam$&quot;)
bam_path &lt;- file.path(data_dir, bam_files)

# parallel processing
parallel::mclapply(seq_along(bam_files), mc.cores = 8, FUN = function(i){
    
    bam_name &lt;- paste0(&quot;coverage.&quot;, gsub(&quot;.bam&quot;,&quot;&quot;,bam_files[i]))
    
    # create coverage
    bam_coverage &lt;- bam2cov(bam_file = bam_path[i], 
                           type = &quot;PAIRED&quot;, MAPQ = 12)
    
    assign(bam_name, bam_coverage)
    
    # save as rda file
    save(list = bam_name, file = paste0(data_dir, bam_name, &quot;.rda&quot;))
    
})

# load coverage files
cov_files &lt;- file.path(data_dir, list.files(path = data_dir, pattern = &quot;.rda$&quot;))

for(i in seq_along(cov_files)){
    load(cov_files[i])
}</code></pre>
<hr />
</div>
<div id="site-centered-matrix" class="section level3">
<h3>Site Centered Matrix</h3>
<ul>
<li>coverages (RleList) can be simply subset by windows surrounding sites</li>
</ul>
<pre class="r"><code>covs &lt;- ls(pattern = &quot;coverage.&quot;)

par(mfcol=c(2,3), cex=1.1)

for(i in seq_along(covs)){
    
    covv &lt;- get(covs[i])
    
    # format name
    main_title &lt;- gsub(&quot;.*E03_|_[G,A,T,C][G,A,T,C].*&quot;, &quot;&quot;, covs[i])
    
    plot(colMeans(as.matrix(covv[sites_4000])), 
         type=&quot;l&quot;, lwd=2, col = &quot;red3&quot;, 
         ylim = c(0,10), xaxt=&quot;n&quot;,
         ylab = &quot;Coverage&quot;, xlab = &quot;Center&quot;,
         main = main_title)
    
    axis(side = 1, at = seq(1,win_size, length.out = 5), 
         labels =  c(-(win_size/2),&quot;&quot;,0,&quot;&quot;,(win_size/2))  )
    
    
}</code></pre>
<p><img src="rmd_BasicChIPseq_files/figure-html/unnamed-chunk-6-1.png" width="960" style="display: block; margin: auto;" /></p>
<hr />
</div>
<div id="input-normalization" class="section level3">
<h3>Input Normalization</h3>
<hr />
</div>
</div>
<div id="data-source" class="section level1">
<h1>Data Source</h1>
<p>The dataset was generated by Nadia Prayitno (LMU, BMC, Becker group)</p>
<hr />
</div>
<div id="references" class="section level1">
<h1>References</h1>
<p>Progressive dosage compensation during Drosophila embryogenesis is reflected by gene arrangement. Prayitno K, Schauer T, Regnard C, Becker PB. EMBO Rep. 2019 Aug;20(8):e48138. doi: 10.15252/embr.201948138. Epub 2019 Jul 9. PMID: 31286660</p>
<p>Note: this pipeline (subsampling etc.) differs from the papers version, but findings are the same either way</p>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
