<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>qPCR - Multigroup Comparison</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="styles.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 60px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h2 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h3 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h4 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h5 {
  padding-top: 65px;
  margin-top: -65px;
}
.section h6 {
  padding-top: 65px;
  margin-top: -65px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Home</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="biostats.html">BioStats</a>
</li>
<li>
  <a href="genomics.html">Genomics</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">qPCR - Multigroup Comparison</h1>

</div>

<div id="TOC">
<ul>
<li><a href="#qpcr-data">qPCR Data</a><ul>
<li><a href="#description">Description</a></li>
<li><a href="#calculate-log2-fold-change">Calculate log2 Fold Change</a></li>
</ul></li>
<li><a href="#data-distribution">Data Distribution</a><ul>
<li><a href="#normality-test">Normality test</a></li>
<li><a href="#variance-homogeneity-test">Variance homogeneity test</a></li>
</ul></li>
<li><a href="#test-for-differences">Test for Differences</a><ul>
<li><a href="#fit-anova-model">Fit Anova model</a></li>
<li><a href="#posthoc-tukey-test">PostHoc Tukey Test</a></li>
</ul></li>
<li><a href="#plots">Plots</a><ul>
<li><a href="#mean-difference">Mean Difference</a></li>
</ul></li>
<li><a href="#references">References</a></li>
<li><a href="#data-source">Data Source</a></li>
</ul>
</div>

<style>
pre {
  overflow-x: auto;
}
pre code {
  word-wrap: normal;
  white-space: pre;
}
a {
  font-size: 16px;
}
</style>
<hr />
<div id="qpcr-data" class="section level1">
<h1>qPCR Data</h1>
<div id="description" class="section level3">
<h3>Description</h3>
<ul>
<li>Comparison of expression changes in chromatin remodelling mutants</li>
<li>Two mutations and double mutant were investigated</li>
<li>One target gene and control gene were measured by RT-qPCR</li>
</ul>
<pre class="r"><code># insert data
df &lt;- data.frame(genotype = c(&quot;WT&quot;, &quot;WT&quot;, &quot;WT&quot;,  &quot;mut1&quot;, &quot;mut1&quot;, &quot;mut1&quot;,  &quot;mut2&quot;, &quot;mut2&quot;, &quot;mut2&quot;,  &quot;doublemut&quot;, &quot;doublemut&quot;, &quot;doublemut&quot;),
                 control_gene = c(143.2632804, 143.2016045, 121.5441674,  120.8509654, 135.2941841, 146.6159225,  147.3569572, 159.5510244, 41.53608903,  112.2483313, 145.9524333, 142.2499418),
                 target_gene = c(0.045530984, 0.029892154, 0.018934757,  0.057283056, 0.053640269, 0.072897732,  0.235321652, 0.2810932, 0.067345158,  0.1904679, 0.271456991, 0.158710149)
)

df</code></pre>
<pre><code>##     genotype control_gene target_gene
## 1         WT    143.26328  0.04553098
## 2         WT    143.20160  0.02989215
## 3         WT    121.54417  0.01893476
## 4       mut1    120.85097  0.05728306
## 5       mut1    135.29418  0.05364027
## 6       mut1    146.61592  0.07289773
## 7       mut2    147.35696  0.23532165
## 8       mut2    159.55102  0.28109320
## 9       mut2     41.53609  0.06734516
## 10 doublemut    112.24833  0.19046790
## 11 doublemut    145.95243  0.27145699
## 12 doublemut    142.24994  0.15871015</code></pre>
</div>
<div id="calculate-log2-fold-change" class="section level3">
<h3>Calculate log2 Fold Change</h3>
<pre class="r"><code># target gene vs control gene
df$diff &lt;- log2(df$target_gene) - log2(df$control_gene)

# either way is the same
cbind(log2(df$target_gene) - log2(df$control_gene),
      log2(df$target_gene / df$control_gene))</code></pre>
<pre><code>##             [,1]       [,2]
##  [1,] -11.619533 -11.619533
##  [2,] -12.225993 -12.225993
##  [3,] -12.648156 -12.648156
##  [4,] -11.042833 -11.042833
##  [5,] -11.300496 -11.300496
##  [6,] -10.973880 -10.973880
##  [7,]  -9.290465  -9.290465
##  [8,]  -9.148754  -9.148754
##  [9,]  -9.268575  -9.268575
## [10,]  -9.202930  -9.202930
## [11,]  -9.070559  -9.070559
## [12,]  -9.807818  -9.807818</code></pre>
<hr />
</div>
</div>
<div id="data-distribution" class="section level1">
<h1>Data Distribution</h1>
<div id="normality-test" class="section level3">
<h3>Normality test</h3>
<pre class="r"><code># test whether it is significantly different from normal distribution
shapiro.test(x = df$diff)</code></pre>
<pre><code>## 
##  Shapiro-Wilk normality test
## 
## data:  df$diff
## W = 0.87117, p-value = 0.06763</code></pre>
<pre class="r"><code># visual inspection
library(ggpubr)
ggqqplot(df$diff)</code></pre>
<p><img src="rmd_qPCR_files/figure-html/unnamed-chunk-3-1.png" width="480" /></p>
</div>
<div id="variance-homogeneity-test" class="section level3">
<h3>Variance homogeneity test</h3>
<pre class="r"><code># test whether group variances are different
library(car)

leveneTest(diff ~ genotype, data = df) </code></pre>
<pre><code>## Levene&#39;s Test for Homogeneity of Variance (center = median)
##       Df F value Pr(&gt;F)
## group  3  0.9722 0.4522
##        8</code></pre>
<hr />
</div>
</div>
<div id="test-for-differences" class="section level1">
<h1>Test for Differences</h1>
<div id="fit-anova-model" class="section level3">
<h3>Fit Anova model</h3>
<pre class="r"><code># if assumptions are met fit anova
fit &lt;- aov(formula = diff ~ genotype, data = df)

summary(fit)</code></pre>
<pre><code>##             Df Sum Sq Mean Sq F value   Pr(&gt;F)    
## genotype     3 18.089   6.030   52.74 1.29e-05 ***
## Residuals    8  0.915   0.114                     
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1</code></pre>
</div>
<div id="posthoc-tukey-test" class="section level3">
<h3>PostHoc Tukey Test</h3>
<pre class="r"><code># if anova test is significant perform multiple pair-wise comparisons
test &lt;- TukeyHSD(fit, ordered = TRUE)

test</code></pre>
<pre><code>##   Tukey multiple comparisons of means
##     95% family-wise confidence level
##     factor levels have been ordered
## 
## Fit: aov(formula = diff ~ genotype, data = df)
## 
## $genotype
##                     diff        lwr      upr     p adj
## mut1-WT        1.0588244  0.1747417 1.942907 0.0207992
## doublemut-WT   2.8041249  1.9200422 3.688208 0.0000352
## mut2-WT        2.9286292  2.0445465 3.812712 0.0000255
## doublemut-mut1 1.7453005  0.8612179 2.629383 0.0010280
## mut2-mut1      1.8698048  0.9857221 2.753887 0.0006448
## mut2-doublemut 0.1245043 -0.7595784 1.008587 0.9675757</code></pre>
<hr />
</div>
</div>
<div id="plots" class="section level1">
<h1>Plots</h1>
<div id="mean-difference" class="section level3">
<h3>Mean Difference</h3>
<pre class="r"><code># plot mean differences with 95% confidence intervals
par(mar=c(4,8,4,1))

plot(test, las=1)</code></pre>
<p><img src="rmd_qPCR_files/figure-html/unnamed-chunk-7-1.png" width="480" /></p>
<ul>
<li>Conclusions: all mutants are significantly different from wild-type. Mutation 2 and double mutant are significantly different from mutation 1 but not from each other. This suggests an epistatic genetic interaction between the two mutation on the target gene.</li>
</ul>
<!-- ### Fold Changes -->
<!-- ```{r, fig.height=5, fig.width=5} -->
<!-- # exponentiate mean and confidence intervals -->
<!-- # error bars are not symmetric -->
<!-- par(mar=c(4,4,4,1)) -->
<!-- centers <- barplot(2^test$genotype[1:3,1],  -->
<!--                    ylim = c(0,16), ylab = "Fold Change") -->
<!-- arrows(x0 = centers,  -->
<!--        x1 = centers,  -->
<!--        y0 = 2^test$genotype[1:3,2],  -->
<!--        y1 = 2^test$genotype[1:3,3],  -->
<!--        angle = 90, code = 3, length = 0.1) -->
<!-- ``` -->
<hr />
</div>
</div>
<div id="references" class="section level1">
<h1>References</h1>
<p><a href="http://www.sthda.com/english/wiki/wiki.php?title=normality-test-in-r" class="uri">http://www.sthda.com/english/wiki/wiki.php?title=normality-test-in-r</a></p>
<p><a href="http://www.sthda.com/english/wiki/compare-multiple-sample-variances-in-r" class="uri">http://www.sthda.com/english/wiki/compare-multiple-sample-variances-in-r</a></p>
<p><a href="http://www.sthda.com/english/wiki/wiki.php?title=one-way-anova-test-in-r" class="uri">http://www.sthda.com/english/wiki/wiki.php?title=one-way-anova-test-in-r</a></p>
<hr />
</div>
<div id="data-source" class="section level1">
<h1>Data Source</h1>
<p>The dataset was provided by Magdalena Murawska (LMU, BMC, Ladurner group)</p>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
